piq/pieapp.py

```diff
@@ -0,0 +1,205 @@
+"""
+Implementation of PieAPP
+References:
+    .. [1] Ekta Prashnani, Hong Cai, Yasamin Mostofi, Pradeep Sen
+    (2018). PieAPP: Perceptual Image-Error Assessment through Pairwise Preference
+    https://arxiv.org/abs/1806.02067
+"""
+from typing import Union, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.modules.loss import _Loss
+
+from piq.utils import _validate_input, _adjust_dimensions
+from piq.functional import crop_patches
+
+
+class PieAPPModel(nn.Module):
+    """Model used for PieAPP score computation
+    Args:
+        num_features: Base feature size, which is multiplied by 2 every 2 blocks
+    """
+    def __init__(self, features=64):
+        super().__init__()
+
+        self.features = features
+        self.pool = nn.MaxPool2d(2, 2)
+        self.flatten = nn.Flatten(start_dim=1)
+
+        self.conv1 = nn.Conv2d(3, features, kernel_size=3, padding=1)
+        self.conv2 = nn.Conv2d(features, features, kernel_size=3, padding=1)
+        self.conv3 = nn.Conv2d(features, features, kernel_size=3, padding=1)
+        self.conv4 = nn.Conv2d(features, features * 2, kernel_size=3, padding=1)
+        self.conv5 = nn.Conv2d(features * 2, features * 2, kernel_size=3, padding=1)
+        self.conv6 = nn.Conv2d(features * 2, features * 2, kernel_size=3, padding=1)
+        self.conv7 = nn.Conv2d(features * 2, features * 4, kernel_size=3, padding=1)
+        self.conv8 = nn.Conv2d(features * 4, features * 4, kernel_size=3, padding=1)
+        self.conv9 = nn.Conv2d(features * 4, features * 4, kernel_size=3, padding=1)
+        self.conv10 = nn.Conv2d(features * 4, features * 8, kernel_size=3, padding=1)
+        self.conv11 = nn.Conv2d(features * 8, features * 8, kernel_size=3, padding=1)
+
+        self.fc1_score = nn.Linear(in_features=120832, out_features=512, bias=True)
```

Comment:
I guess this line makes having `features` as a parameter senseless because number of input features is hardcoded. I also think that since you use pretrained model weights you can't really change parameters.

Reply:
Agree. Removed it from __init__ and hardcoded for now.
As soon as dataset used for PieAPP model training will be released we can reconsider this implementation

Reply:
I also added a TODO comment to not forget